### ISSUE TITLE
funding: use mempool.MaxRBFSequence for funding transactions we create

### ISSUE THREAD

**Issue by Roasbeef** [lightningnetwork/lnd#8064]
We should use a sequence of `mempool.MaxRBFSequence` for funding transactions we create. For the new musig2 based channel type, we already use this for the co-op close transactions, so doing it for funding transactions would help things blend in a bit more. 

This'll also make it easier to add an automated way to double spend a funding transaction (before it hits the chain), if a user messes up with something like the custom channel funding (have had instances of users using Electrum to RBF, causing the channel to be lost). This makes it easier to mess with zero conf channels, but that's already a trusted feature, and our implementation forces the acceptor to explicitly accept it with the channel acceptor.

FWIW, those that use the custom or PSBT based funding can already do this as well.

**Comment by ziggie1984** [comment]
I think is also makes sense to use the `mempool.MaxRBFSequence` when sending coins from the wallet. Currently it creates transaction which signal no RBF, so would be great to have all transactions created by lnd to signal RBF.

For example:
`lncli sendcoins`

**Comment by madskrogh** [comment]
@ziggie1984 i'm looking into this issue and finished the funding tx part. 

But as for `lncli sendcoins` i believe it gets an already signed transaction from btcwallet  which means the changes have to be made there. https://github.com/lightningnetwork/lnd/blob/fad7d3fe8188809914cd5fff65c6de8a53f70f01/lnwallet/btcwallet/btcwallet.go#L1029
If that's okay i'll bundle my LND pr with one for btcsuite/btcwallet – just wanted to check first.

**Comment by ziggie1984** [comment]
Thanks for picking this up @madskrogh. Yes I think it makes sense to have one PR which touches both of them. `sendcoins` was just an example I think `fundpsbt` is part of this too. Maybe check which commands could also increase our privacy if the would signal RBF?

**Comment by madskrogh** [comment]
@ziggie1984 I have now implemented RBF for the following commands:
- [x] `openchannel` (including with `sweepall` flag)
- [x] `sendcoins`
- [x] `sendmany`
- [x] `psbt fund`
- [x] `batchopenchannel`

How about HTLC and anchor sweep txs?

**Comment by ziggie1984** [comment]
Cool, no the sweeper should not be changed, because there we cannot just default the sequence number (we do have different CSVs). The privacy for Force-Closes is very weak anyways because all Scripts are published onchain.
You can start opening PR's (lnd and btcwallet).

**Comment by madskrogh** [comment]
Quick thought before i submit the prs: why only RBF for the co-op close transactions of taproot channels (as stated in the original issue description)? Do we not want other older channel types to co-op close with RBF too?

**Comment by ziggie1984** [comment]
That's a spec change and because we do reveal the 2of2 multisig script (not the case for taproot) I think it is not worth to change and I am not even sure if it would be a win, I don't think so.

https://github.com/lightning/bolts/blob/master/03-transactions.md#closing-transaction

**Comment by guggero** [comment]
Haven't taken a look at the PRs yet, but I think we should _NOT_ signal RBF for any channel funding transactions! Replacing a channel funding transaction invalidates any pre-signed commitment transactions, which can lead to loss of funds. So even though a technical user can still create a replacement transaction and broadcast it (which unfortunately happens from time to time), we shouldn't put an explicit signal into the transaction itself.

**Comment by morehouse** [comment]
> Haven't taken a look at the PRs yet, but I think we should _NOT_ signal RBF for any channel funding transactions! Replacing a channel funding transaction invalidates any pre-signed commitment transactions, which can lead to loss of funds. So even though a technical user can still create a replacement transaction and broadcast it (which unfortunately happens from time to time), we shouldn't put an explicit signal into the transaction itself.

If the funding tx is replaced, the funding outpoint will change, so neither side will consider the channel to have opened.  Unless there's a major bug of some sort, I don't see how we could lose funds.

**Comment by guggero** [comment]
Well, if you RBF the transaction and still send coins to the 2-of-2 p2wsh channel funding output (instead of sending the coins to another wallet address), then you send funds to a 2-of-2 multisig to which your node only has a single key and no valid commitment transaction to spend again. This is what users unfortunately do today sometimes (see https://github.com/lightningnetwork/lnd/discussions/8208 for example). So making this worse by explicitly signaling (signalling here meaning more to the user than to the network) RBF on channel funding transactions IMO isn't a good idea.

**Comment by morehouse** [comment]
Ah, I see.  The risk is the user shooting themself in the foot.

I still think there's good reasons to signal RBF.

- better privacy for taproot channels
- ability to cancel channel opens
- interactive-tx (dual funding) requires signalling RBF

I think RBF is the direction we want to head in.  The question is how to better warn users about the footgun.

**Comment by guggero** [comment]
Okay, those are pretty good arguments for using RBF indeed, especially the privacy one for Taproot Channels.
Not sure how we can best prevent users from shooting themselves in the foot. Or whether a TX showing up as "RBF" on a block explorer would actually give them the idea to manually attempt to replace it...

**Comment by ziggie1984** [comment]
It would be interesting to know, how the user was able to sign the new RBF transaction, probably he used some psbt signing function of lnd ? Maybe we can build in a check there that we verify the channel-openings as soon as we sign a wallet transaction und be sure not the same P2WSH/P2TR address is used, and log a nice warning when the user tries so?

**Comment by guggero** [comment]
No, at least from two instances I know they imported the keys into Electrum which offers the RBF functionality. Apparently there is a guide out there somewhere that explains how to do that, though I couldn't find it. I'm pretty sure it's hard to pull off with just the `lnd` APIs, so I don't think we can do a lot on our side.

**Comment by morehouse** [comment]
> I'm pretty sure it's hard to pull off with just the `lnd` APIs, so I don't think we can do a lot on our side.

Perhaps we can better advertise the [proper way](https://docs.lightning.engineering/lightning-network-tools/lnd/unconfirmed-bitcoin-transactions#docs-internal-guid-d7a5497d-7fff-d4c9-e294-cb0cec6c9e86) to fee bump the funding tx.  Ideally, we provide such an intuitive way to fee bump that the user has no reason to search for more complicated things.

There may be some usability improvements we could make to the existing API.  For example, we could detect that a funding tx was specified and by default automatically choose the next feerate and the output for CPFP:

```shell
lncli bumpfee 5f97234af4df23881ca5e994bf42956c9af53bf83cb63687d1d0adfcd3ece18a
```

In the future, if we have v2 channel opens, the same API could initiate an RBF negotiation with the peer automatically.

**Comment by ziggie1984** [comment]
> There may be some usability improvements we could make to the existing API. For example, we could detect that a funding tx was specified and by default automatically choose the next feerate and the output for CPFP:

I think it's a good idea  or even introduce a new command to separate the logic between the sweeper feebumping and the channelopening ? 

The new API would do the following:

1. Check whether we are the opener of the channel and if a change_output exists for us to bump, calculate the exact parent+child weight to bump with the exact feerate (effective feerate).
2. If a CPFP already exist make sure it does not have any descendants before RBFing it to higher feeRate, if a descendant exist look for another change_output on the opening tx and if not signal a warning that further FeeBumping is not possible (because we cannot be sure whether this output was used for zero_conf channel openings)?
3. If we are not the opener or there does not exist a change output show a WARNING that RBFing this transaction and creating a new transaction paying to the MultiSig address will lead to loss of funds.

Having a new API let's the user clearly see that lnd allows for bumping the channel opening and if they use it they will likely see the warning when using it.

So it makes sense to only signal RBF for the opening of a channel once this API is ready or ?
